gapis/api/watcher.go

```diff
@@ -0,0 +1,94 @@
+// Copyright (C) 2018 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package api
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/google/gapid/gapis/memory"
+)
+
+type Field interface {
+	FieldName() string
+	ClassName() string
+}
+
+// StateWatcher provides callbacks to track state effects
+type StateWatcher interface {
+	// OnBeginCmd is called at the beginning of each API call
+	OnBeginCmd(ctx context.Context, cmdID CmdID, cmd Cmd)
+
+	// OnEndCmd is called by the API emulator at the end of each API call
+	OnEndCmd(ctx context.Context, cmdID CmdID, cmd Cmd)
+
+	// OnGet is called when a fragment of state (field, map key, array index) is read
+	OnGet(ctx context.Context, owner Reference, f Fragment, v Reference)
+
+	// OnSet is called when a fragment of state (field, map key, array index) is written
+	OnSet(ctx context.Context, owner Reference, f Fragment, old Reference, new Reference)
+
+	// OnWriteSlice is called when writing to a slice
+	OnWriteSlice(ctx context.Context, s memory.Slice)
+
+	// OnReadSlice is called when reading from a slice
+	OnReadSlice(ctx context.Context, s memory.Slice)
+
+	// OnWriteObs is called when a memory write observation becomes visible
+	OnWriteObs(ctx context.Context, index int, obs CmdObservation)
```

Comment:
For performance reasons, would it make sense to pass all the observations in this call?

Reply:
Done. I don't think it will have a significant performance impact, but it does make the interface cleaner by removing the `index` parameter.